Feat/nix shell

ESCristiano opened the pull request:
## PR Description

Add a nix-shell environment to contemplate all essential bao-project dependencies, allowing developers to effortlessly work across all bao-project repositories.

The user can utilize the bao-project nix-shell in two ways:  

1. Run a standalone command within the bao-project nix-shell environment;

```bash
make nix-shell-run cmd="your command"
```

2. Open an interactive bao-project nix-shell environment.

```bash
make nix-shell
```
**Supported Bao-Project Repositories**

The ultimate goal is to provide support for all the bao-projects repositories. Currently, the nix-shell only handles dependencies for bao-docs. Support for the remaining repositories will be incorporated in the future.

### Type of change

- **feat**: introduces a new functionality
  - Logical unit: `shell.nix` 

## Checklist:

- [x] The changes generate no new warnings when building the project. If so, I have justified above.
- [x] I have run the CI checkers before submitting the PR to avoid unnecessary runs of the workflow.

AfonsoSantos96 commented on ci.mk:
```diff
@@ -340,5 +340,32 @@ endef
+#    make nix-shell
+# To run a standalone commnand inside the nix-shell use:
+#    make nix-shell-run cmd="command-to-run"
+# @param The command to run inside the nix-shell.
```
`@param command-to-run The command to run inside the nix-shell.`
I understood the description but it is in order to obey the doxygen comment style `@param <variable_name> <description>`

ESCristiano replied:
I understand your comment, however, I just followed the description used in other rules on the makefile.

josecm commented on README.md:
```diff
@@ -139,6 +138,38 @@ locally.
 
 ---
 
+## Using the Bao-Project Nix Shell
+
+Nix provides a robust and reproducible environment by encapsulating all necessary dependencies within a controlled environment. This approach effectively eliminates issues associated with version conflicts and missing packages/dependencies, ensuring a seamless and reliable development experience.
+
+We offer a nix-shell environment containing all essential Bao-Project dependencies, allowing developers to effortlessly work across all Bao-Project repositories. The required dependencies will be automatically fetched. If you already have the necessary packages/dependencies installed, Nix will prioritize the ones it fetched, ensuring reproducibility. Once you exit the nix-shell environment, those ...
+
+You can utilize the Bao-Project nix-shell in two ways:  (1) run a standalone command within the Bao-Project nix-shell environment; (2) or open an interactive Bao-Project nix-shell environment.
+
+**Running a Standalone Command**
+
+This command opens a nix-shell, executes the provided user command, and automatically closes the shell.
+
```
Please limit the column width to 80 characters.

ESCristiano replied:
Well spotted, I will add a reference to [Nix official installation](https://nixos.org/download.html) page.

josecm commented on README.md:
````diff
@@ -139,6 +138,38 @@ locally.
+
+You can utilize the Bao-Project nix-shell in two ways:  (1) run a standalone command within the Bao-Project nix-shell environment; (2) or open an interactive Bao-Project nix-shell environment.
+
+**Running a Standalone Command**
+
+This command opens a nix-shell, executes the provided user command, and automatically closes the shell.
+
+```bash
+make nix-shell-run cmd="your command"
+```
+
+**Opening a Nix-Shell Environment**
+
+This command opens an interactive nix-shell with all the Bao-Project dependencies. 
+
+```bash
+make nix-shell
````
I was thinking we could follow the same logic as for docker. Having a directory named "nix", with a fully separate Makefile. Then, running make commands in that directory would automatically execute it inside the nix-shell. For example, from a project's top directory, running:

```
make -C ci/nix format-check
```

would run the command `make format-check` command in the nix-shell. This is the same logic we use for docker.

In that makefile we can have other targets for running arbitrary commands and starting the interactive shell as you implemented in the ci.mk file.

ESCristiano replied:
As I mentioned above, I will implement your suggestion. It also makes sense to me.

josecm reviewed (changes requested):
We should also modify the README introduction to highlight we have this build environment two.

josecm commented:
@ESCristiano I've updated the README introduction to mention the nix-shell facility.

ESCristiano commented on ci.mk:
```diff
@@ -340,5 +340,32 @@ endef
+# bao-project repositories. 
+# To open an interactive nix-shell with the bao-project environment:
+#    make nix-shell
+# To run a standalone commnand inside the nix-shell use:
```
I will fix the typo.

ESCristiano commented on ci.mk:
```diff
@@ -340,5 +340,32 @@ endef
 
 #############################################################################
 
+# Nix-shell
```
It makes perfect sense to me as well. In fact, while implementing the Nix functionality, I found myself wondering the same thing. I will implement your suggestion.

ESCristiano commented:
@josecm I've incorporated your suggestion and now nix can be used in a similar way as docker. 

I've also updated the PR description to reflect these changes.

josecm commented on nix/Makefile:
```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) Bao Project and Contributors. All rights reserved
+
+root_dir?=$(realpath ..)
```
I'd say the root dir by default should be:

```suggestion
root_dir?=$(realpath ../../)
```

Which, by our convention on how the ci repo should be included in the project's repos, should be the repo top-level directory and where all commands by default should run. And then, in the recipes:

```makefile
nix-shell:
	@cd $(root_dir) && nix-shell $(nix_dir)/shell.nix

nix-shell-run:
	@cd $(root_dir) && nix-shell $(nix_dir)/shell.nix --run "$(cmd)"
```

josecm commented on ci.mk:
Since we have the nix only makefile, I think the modifications for this file should be removed.

josecm commented on nix/Makefile:
```diff
@@ -0,0 +1,40 @@
+	@nix-shell $(nix_dir)/shell.nix --run "$(cmd)"
+
+.PHONY: all
+non_build_targets+=nix-shell nix-shell-run
+
+define nix-shell-run
+cmd:=$1
+endef
```
I don't think these is actually needed.
